feat(rls): Add Groups association to RLS

Antonio-RiveroMartnez opened the pull request:
### SUMMARY
We recently introduced the `Groups` [with this PR](https://github.com/apache/superset/pull/32121), these changes are adding support for them on the `RLS` rules. This way it become even more friendly and usable setting rules for users. 

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
#### Creating a RLS rule

##### Regular rule
<img width="481" alt="Screenshot 2025-03-18 at 11 47 14 AM" src="https://github.com/user-attachments/assets/44245af3-62b2-419f-93f9-590c1378b003" />

##### Base rule
<img width="481" alt="Screenshot 2025-03-18 at 11 47 26 AM" src="https://github.com/user-attachments/assets/0e046584-510b-4e10-9f1f-29cda90d57b6" />

#### Rules being applied based on groups from direct user <-> group association and rule being linked to the group
<img width="1648" alt="Screenshot 2025-03-18 at 11 54 34 AM" src="https://github.com/user-attachments/assets/6139354f-66a9-4854-a6ec-4d3efd07e8a8" />

#### Rules being applied based on groups from role association (i.e, no direct user <-> group relation, but the group linked to a role and the user linked to that same role)

##### RLS rule details
<img width="480" alt="Screenshot 2025-03-18 at 11 58 38 AM" src="https://github.com/user-attachments/assets/6728d29a-b194-4334-89a9-775064f68fff" />

#####  Group details
<img width="480" alt="Screenshot 2025-03-18 at 11 58 53 AM" src="https://github.com/user-attachments/assets/c6b2a151-5943-457e-9b90-13879084f697" />

##### Rule applied
<img width="1660" alt="Screenshot 2025-03-18 at 11 59 52 AM" src="https://github.com/user-attachments/assets/e445c110-62a4-4f1a-b27f-5527421475f6" />



### TESTING INSTRUCTIONS
1. Create group(s) using the existing functionality
2. Access the Row Level Security menu and associate the group(s) you created
3. Verify the clause gets into the generated query when for example creating a chart using a restricted table

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [ ] Has associated issue:
- [ ] Required feature flags:
- [X] Changes UI
- [X] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [X] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

Antonio-RiveroMartnez committed:
feat(rls): Add groups to RLS

- Add new table to hold many to many relation between rls and groups
- Support groups as part of rls APIs
- Consider groups when fetching rls filters
- Update the UI to allow group management for rls

Antonio-RiveroMartnez committed:
- Minor adjustments to API fields and specification
- Tests for UI changes

Antonio-RiveroMartnez committed:
- Create and use new custom exception in RLS APIs when groupd is not valid
- Move group query to util method and use it from commands
- Adjust schemas for API
- Consider resolving RLS clauses with roles coming from groups and viceversa
- Add integration tests to our RLS APIs

Antonio-RiveroMartnez committed:
- Add RLS logic tests

Antonio-RiveroMartnez committed:
- fixing ruff checks
- Fix lint on ui test

Antonio-RiveroMartnez committed:
- Adding constraints to new table and model

dpgaspar commented on superset/connectors/sqla/models.py:
```diff
@@ -2099,6 +2099,25 @@ def load_database(self: SqlaTable) -> None:
+        ),
+    ),
+    UniqueConstraint("group_id", "rls_filter_id", name="uq_rls_filter_group"),
+)
```
for roles we have:
```
RLSFilterRoles = DBTable(
    "rls_filter_roles",
    metadata,
    Column("id", Integer, primary_key=True),
    Column("role_id", Integer, ForeignKey("ab_role.id"), nullable=False),
    Column("rls_filter_id", Integer, ForeignKey("row_level_security_filters.id")),
)
```
do we need to remove `nullable=False` now? because we can have a roles or a group?

dpgaspar commented on superset/security/manager.py:
```diff
@@ -2473,32 +2481,73 @@ def get_rls_filters(self, table: "BaseDatasource") -> list[SqlaQuery]:
             return []
 
         # pylint: disable=import-outside-toplevel
+        from flask_appbuilder.security.sqla.models import assoc_group_role
```
nit: we could move this one to a top level import

dpgaspar commented on superset/security/manager.py:
```diff
@@ -2473,32 +2481,73 @@ def get_rls_filters(self, table: "BaseDatasource") -> list[SqlaQuery]:
         )
 
+        # Roles and Groups directly linked to the user
         user_roles = [role.id for role in self.get_user_roles(g.user)]
```
`get_user_roles(g.user)` will get all roles that are assigned to a user, directly or indirectly via groups

Antonio-RiveroMartnez committed:
- Make group_id not nullable in the assoc table
- Remove duplicate import
- Leverage get_user_roles that already gives you all the roles of a user no matter if direct or indirect (via groups)

villebro commented:
@Antonio-RiveroMartnez @dpgaspar as proposed in [SIP-126](https://github.com/apache/superset/issues/28021), which was integrated into [SIP-152](https://github.com/apache/superset/issues/32116), there was talk of introducing a Subject model that could be either a Role, Group or User. As we start rolling out Groups into the various entity models, it feels introducing that new model + component now would be very beneficial, as otherwise we'll have to duplicate the components (one for roles, one for groups and later one for users). This will become particularly convoluted in the Dashboard and Chart models, if we introduce editors and viewers, in which case there would be 6 dropdowns all together. Doing that work now would avoid making big design overhauls later, and would save development time by not having to migrate to the new model when it's ultimately introduced.

Thoughts?

Antonio-RiveroMartnez commented:
Hey @villebro @michael-s-molina, totally agree on bringing the `Subject` model soon 🙏  but seems to me that it was considered as part of a phase 2 in [SIP-152](https://github.com/apache/superset/issues/32116)? While the `Group` entity in RLS would be still considered phase 1 since it's just a complementary element of the initial PR thus it could potentially be added now? What do you guys think? Thanks for looking into this

michael-s-molina commented:
I think it will be problematic to introduce another select for group on all screens and later consolidate all selects into one. Why don't we start with a single select given that this is a new feature? In other words, why introduce technical debt when we can actually make the experience simpler by reducing the number of things users need to handle? As @villebro pointed out, with the concepts of Editor and Viewer, we would be talking about 6 selects to handle access.

villebro commented:
@Antonio-RiveroMartnez I think it was my oversight not to add RLS as part of the Subject model overhaul (I wrote the [bullet list](https://github.com/apache/superset/issues/32116#issuecomment-2643567384) of entities that will be changed that was later merged into the final SIP description). I totally assumed RSL to be part of this work, as the current Role reference in RLS really is a subject reference.

The distinction in the phasing here IMO is that Phase 1 was all about introducing Groups, and Phase 2 is all about replacing existing subject-like references with the new Subject model. The fact that we're here essentially simulating the new Subject model via having double drop downs is IMO a strong indication that the Subject model + component should be introduced first, as otherwise this change would need to be completely redone in Phase 2.

Antonio-RiveroMartnez commented:
@villebro @michael-s-molina If `Groups` are meant to be introduced in `Phase 1`, shouldn’t they also work within RLS to be fully integrated? otherwise, they are partially usable. Also, as things stand, a refactor will be needed when we introduce the `Subject` entity regardless of `Groups` being part of `RLS` or not, but I see that we would be adding extra work on top of that.

I completely understand the concerns you've raised and I truly appreciate your time to review and having this discussion, I'm just trying to fully grasp the `Phase 1 vs Phase 2` approach so I can contribute in the best way possible.

Thanks.

dpgaspar commented:
> @Antonio-RiveroMartnez I think it was my oversight not to add RLS as part of the Subject model overhaul (I wrote the [bullet list](https://github.com/apache/superset/issues/32116#issuecomment-2643567384) of entities that will be changed that was later merged into the final SIP description). I totally assumed RLS to be part of this work, as the current Role reference in RLS really is a subject reference.
> 
> The distinction in the phasing here IMO is that Phase 1 is all about introducing Groups, and Phase 2 is all about replacing existing subject-like references with the new Subject model. The fact that we're here essentially simulating the new Subject model via double drop downs is IMO a strong indication that the Subject model + component should be introduced first, as otherwise this change would need to be completely redone in Phase 2.

@villebro @michael-s-molina I understand the concerns around aligning RLS with the Subject model. However, as @Antonio-RiveroMartnez pointed out, I also firmly believe that RLS should be introduced in Phase 1, without being tied to the Subject model yet. The core goal of Phase 1 is to introduce Groups in a backward-compatible way, and leaving them out of RLS at this stage would result in an incomplete and inconsistent implementation.

Per SIP-152, Phase 1 must leave Superset in a fully functional state with respect to API and datasource authorization, ensuring a smooth transition while we prepare for the breaking changes in Phase 2. Some refactoring will be necessary in Phase 1, but that would be the case regardless of RLS.

I don’t think adding a new dropdown to RLS is a strong enough reason to either prematurely introduce the Subject model or leave Phase 1 incomplete. Is it a perfect solution? No. But major changes rarely follow a straight, clean path.

villebro commented:
> Per SIP-152, Phase 1 must leave Superset in a fully functional state with respect to API and datasource authorization, ensuring a smooth transition while we prepare for the breaking changes in Phase 2. Some refactoring will be necessary in Phase 1, but that would be the case regardless of RLS.

@dpgaspar I'm probably missing something, as I assumed RLS should work as before without this change. Is there something in datasource authorization that has changed with the introduction of groups, causing issues for RLS? If this is indeed the case, then I can understand this intermediate change. But if not, I still feel this change is best suited for Phase 2 to avoid introducing this temporary workaround.

eschutho commented:
@michael-s-molina @villebro what do you think about us creating a subject dropdown that allows people to select either a group or a role? It will be an intermediary step closer to subjects, and then we wouldn't need to have two dropdowns.

villebro commented:
> @michael-s-molina @villebro what do you think about us creating a subject dropdown that allows people to select either a group or a role? It will be an intermediary step closer to subjects, and then we wouldn't need to have two dropdowns.

That would make it impossible to mix and match different subject types. I also assume it would require an extra db migration, by first adding a group reference to RLS, and later replacing that with the final subject reference (likely a new table).

dpgaspar commented:
> > @michael-s-molina @villebro what do you think about us creating a subject dropdown that allows people to select either a group or a role? It will be an intermediary step closer to subjects, and then we wouldn't need to have two dropdowns.
> 
> That would make it impossible to mix and match different subject types. I also assume it would require an extra db migration, by first adding a group reference to RLS, and later replacing that with the final subject reference (likely a new table).

Yes, it would probably require 2 migrations to get to our final destination (depends on the specifics of the implementation itself). But on the bright side, it would fully introduce `Groups` to Superset, that's 1 step forward to subjects, also I don't believe subjects implementation is right around the corner for multiple reasons, and I do think it's positive to get Superset to a state where `Groups` is a first class entity.
@eschutho proposal, could be a good compromise because, it would not introduce a new field to the RLS form addressing @michael-s-molina concern here: https://github.com/apache/superset/pull/32718#issuecomment-2734060623

Another alternative would be to feature flag this dropdown, not my first choice.

Our goal here is to get Phase 1 to be a stable and complete point regarding data authorization and RLS is an important feature on data authorization. Our users can't fully adopt `Groups` without RLS.

I also believe that introducing subject models and all it requires at this point is premature and would risk a refactor

I do agree and I'm fully onboard on stoping unnecessary tech debt, but I think this change is manageable and required, it's a compromise, also the next step is planed and approved on [SIP-152](https://github.com/apache/superset/issues/32116) that incorporated [SIP-126](https://github.com/apache/superset/issues/28021)

Yes, implementing this would likely require two migrations to reach our final goal, though the specifics depend on the implementation. However, on the positive side, this approach would fully introduce Groups into Superset, a step toward the broader Subjects concept. Given that a full `Subjects` implementation isn't imminent for several reasons, making Groups a first-class entity now makes sense (and was approved).

@eschutho's proposal seems like a reasonable compromise, as it avoids introducing a new field to the RLS form—directly addressing @michael-s-molina's concern [here](https://github.com/apache/superset/pull/32718#issuecomment-2734060623). Another alternative could be feature-flagging the dropdown, though that wouldn't be my first choice.

Our primary goal is to ensure Phase 1 is a stable and complete step forward in data authorization, with RLS playing a crucial role. Our users can't fully adopt `Groups` without RLS, making this change both necessary and timely.

Introducing now all the necessary backend implementation for Subjects is premature and could risk a refactor, since 

I agree that we should avoid unnecessary tech debt, but I believe this change is small contained and justified. It strikes a balance between progress and stability, especially considering that the next step has already been planned and approved in [SIP-152](https://github.com/apache/superset/issues/32116), which builds upon [SIP-126](https://github.com/apache/superset/issues/28021).

Unless there are strong objections, I think this is a reasonable path forward.

Antonio-RiveroMartnez commented:
Closing this PR in favor of the `Subjects` entity coming in `Phase 2`
